src/Client/Updater.php

```diff
@@ -528,7 +529,47 @@ private static function hasRotatedKeys(RootMetadata $previousRootData, RootMetad
      */
     private function fetchFile(string $fileName, int $maxBytes = self::MAXIMUM_DOWNLOAD_BYTES): string
     {
-        return $this->repoFileFetcher->fetchMetaData($fileName, $maxBytes)->wait();
+        return $this->repoFileFetcher->fetchMetaData($fileName, $maxBytes)
+            ->then(function (StreamInterface $data) use ($fileName, $maxBytes) {
+                $this->checkLength($data, $maxBytes, $fileName);
+                return $data;
+            })
+            ->wait();
+    }
+
+    /**
+     * Verifies the length of a data stream.
+     *
+     * @param \Psr\Http\Message\StreamInterface $data
+     *   The data stream to check.
+     * @param int $maxBytes
+     *   The maximum acceptable length of the stream, in bytes.
+     * @param string $fileName
+     *   The filename associated with the stream.
+     *
+     * @throws \Tuf\Exception\DownloadSizeException
+     *   If the stream's length exceeds $maxBytes in size.
+     */
+    private function checkLength(StreamInterface $data, int $maxBytes, string $fileName): void
+    {
+        $error = new DownloadSizeException("$fileName exceeded $maxBytes bytes");
+        $size = $data->getSize();
+
+        if (isset($size)) {
+            if ($size > $maxBytes) {
+                throw $error;
+            }
+        } else {
+            $data->read($maxBytes);
+
+            // If we reached the end of the stream, we didn't exceed the
+            // maximum number of bytes.
+            if ($data->eof() === false) {
+                throw $error;
+            }
+            // @todo Handle non-seekable streams with unknown size.
```

Comment:
Is there a way we can tell if a stream is non-seekable here? We could just throw an exception here and say this type of stream is unsupported for now.

Reply:
The docs of `StreamInterface::rewind()` specify that non-seekable streams will throw an exception when that method is called, so we don't need to do anything additional here. But in the future, it would be nice to handle non-seekable streams gracefully if we can.

Reply:
Also, `StreamInterface` has an `isSeekable()` method, precisely for this kind of introspection.

Reply:
Should we call `rewind()` before `read($maxBytes)` to ensure we are at the beginning before test to see if it gets us to the end of the stream. We would still have to call `rewind()` after also.

Reply:
Yeah, that makes sense. I'll add it.